Update EIP-4974: Community Feedback (#5135)
ethereum/EIPs: EIPS/eip-4974.md

---
eip: 4974

> title: Comparable Non-Transferable Tokens

Reviewer: For some reason the Github doesn't let me add the following suggestion to the Line 34.

Fixes the links to EIPs

```
--- [ERC-20](./eip-20) and [ERC-721](./eip-721)
+++ [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md)
```

Author: Thanks, I'll update it.

> description: A standard interface for ratings.

Reviewer 2: I do not believe that "Rating" is the best description for this. How about "Fungible Non-Transferrable Tokens" (but dropping the EXP)?

Reviewer 3: Fungibility as a choice of terminology since it describes "interchangeability." But how can there be fungibility if the tokens cannot be traded?

Reviewer 2: Interchangeability refers to equality between tokens. Basically, it means you can stick a number on some quantity and say "I have X amount of Y," instead of saying "I have the Y." For example, if you have 3 apples, you can say "I have 3 apples," but you can't say "I have the apples." You do not have all the apples in the world (unless there is a _really_ bad shortage of them).

Likewise, the description of this standard means it is fungible. You have a certain _amount_ of the non-tradeable token, not just zero or one. This makes it fungible.

Reviewer 3: I'm no economist but I had understood fungibility to be a trait of the goods exchange or substitution quality. The definition is that you can use either this or that gold/oil because it doesn't matter: it's fungible - individual units of itself can be used as substitution interchangably.

I wanna say that I agree that e.g. the experience points players gain in an MMORPG are comparable. It's not that everyone gains experience points on their personal scale. Still my interpretation is that fungibility only exists within the world of trading.

I know "fungible" is in everyone's hearts and minds. But e.g. the more accurate description IMO would be "Comparable Non-Transferable Tokens"

Reviewer 2: I'm fine with "Comparable Non-Transferable Tokens".

Author: Makes sense, but the focus of this EIP is sharing. It aims to be a standard for one contract to rate another contract or wallet, so I don't think the token terminology is best anymore.

> title: Rating
> description: A standard interface for ratings.

Reviewer 2: ```suggestion
title: Comparable Non-Transferable Tokens
description: ...
```
Not sure what a good description for this would be that doesn't duplicate the title. It can't include the word `standard` or similar words.

---

## Abstract

> The EIP-4974 standard defines a standardized interface for assigning and managing numerical ratings on the Ethereum blockchain. This allows ratings to be codified within smart contracts and recognized by other applications, enabling a wide range of new use cases for tokens.

Reviewer 4: ```suggestion
This standard defines a standardized interface for assigning and managing numerical ratings on the Ethereum blockchain. This allows ratings to be codified within smart contracts and recognized by other applications, enabling a wide range of new use cases for tokens.
```

## Motivation
Traditionally, blockchain applications have focused on buying and selling digital assets. However, the asset-centric model has often been detrimental to community-based blockchain projects, as seen in the pay-to-play dynamics of many EVM-based games and DAOs in 2021.

> EIP-4974 addresses this issue by allowing ratings to be assigned to contracts and wallets, providing a new composable primitive for blockchain applications. This allows for a diverse array of new use cases, such as:

Reviewer 4: ```suggestion
This proposal addresses this issue by allowing ratings to be assigned to contracts and wallets, providing a new composable primitive for blockchain applications. This allows for a diverse array of new use cases, such as:
```

---

## Abstract

> The following describes a standard interface for assignment of numerical ratings. This standard provides basic functionality for an operator address to assign and update ratings.

Reviewer 2: ```suggestion
The following describes a standard interface for the assignment of numerical ratings. This standard provides basic functionality for an operator address to assign and update ratings.
```

> Ratings may represent accumulated recognition within a smart contract. Like experience points in video games, citations on an academic paper, or Reddit Karma, EXP is bestowed for useful contributions, accumulates as indistinguishable units, and should only be reallocated or destroyed by a reliable authority so empowered.

Reviewer 5: You are removing the blurb about what exp point are and not including an explanation in the new line.

Author: Thanks for the note. Fully updated now.

///  Must initialize contracts assigning participation as `true` for both `operator` and `address(0)`.
interface IERC4974 /* is ERC165 */ {

    /// @dev Emits when operator issues a rating. 
    ///  MUST emit when rating is assigned by any mechanism.
    ///  MUST ONLY emit by `rate`.

>     event Rating(address _rated, int8 _rating_);

Reviewer 2: Would be changed to BalanceUpdated

    /// @notice Rate an address.
    ///  MUST emit a Rating event with each successful call.
    /// @param _rated Address to be rated.
    /// @param _rating Total EXP tokens to reallocate.

>     function rate(address _rated, int8 _rating_) external;

Reviewer 2: Would be changed to setBalance

>     function rate(address _rated, int8 _rating_) external;

Author: This is on purpose! 

- It's signed:  Reviewers should be able to give neutral and negative ratings for the wallets and contracts they interact with.
- It's 8bit:  The objective here is to keep ratings within some fathomably comparable range. Longer term, this could encourage easy aggregation of ratings, versus using larger numbers where users might employ a great variety of scales.

Does that make sense?

    /// @notice Return a rated address' rating.
    /// @dev MUST register each time `Rating` emits.
    ///  SHOULD throw for queries about the zero address.
    /// @param _rated An address for whom to query EXP total.
    /// @return int8 The rating assigned.

>     function ratingOf(address _rated) external view returns (uint256);

Reviewer 2: Would be reverted back to balanceOf
